lib/hocon/impl/config_delayed_merge.rb

```diff
@@ -35,10 +46,189 @@ def unwrapped
+        source_for_end = source.replace_within_current_parent(replaceable, remainder)
+
+        if ConfigImpl.trace_substitution_enabled
+          ConfigImpl.trace("  sourceForEnd before reset parents but after replace: #{source_for_end}", new_context.depth)
+        end
+
+        source_for_end = source_for_end.reset_parents
+      else
+        if ConfigImpl.trace_substitution_enabled
+          ConfigImpl.trace("will resolve end against the original source with parent pushed",
+                           new_context.depth)
+        end
+
+        source_for_end = source.push_parent(replaceable)
+      end
+
+      if ConfigImpl.trace_substitution_enabled
+        ConfigImpl.trace("sourceForEnd      =#{source_for_end}", new_context.depth)
+      end
+
+      if ConfigImpl.trace_substitution_enabled
+        ConfigImpl.trace("Resolving highest-priority item in delayed merge #{stack_end}" +
+                          " against #{source_for_end} endWasRemoved=#{(source != sourceForEnd)}")
+      end
+
+      result = new_context.resolve(stack_end, source_for_end)
+      resolved_end = result.value
+      new_context = result.context
+
+      if ! resolved_end.nil?
+        if merged.nil?
+          merged = resolved_end
+        else
+          if ConfigImpl.trace_substitution_enabled
+            ConfigImpl.trace("merging #{merged} with fallback #{resolvedEnd}",
+                             new_context.depth + 1)
+          end
+          merged = merged.with_fallback(resolved_end)
+        end
+      end
+
+      count += 1
+
+      if ConfigImpl.trace_substitution_enabled
+        ConfigImpl.trace("stack merged, yielding: #{merged}",
+                         new_context.depth)
+      end
+    end
+
+    ResolveResult.make(new_context, merged)
+  end
+
+  def make_replacement(context, skipping)
+    self.class.make_replacement(context, stack, skipping)
+  end
+
+  # static method also used by ConfigDelayedMergeObject; end may be null
+  def self.make_replacement(context, stack, skipping)
+    sub_stack = stack.sub_list(skipping, stack.size)
+
+    if sub_stack.empty?
+      if ConfigImpl.trace_substitution_enabled
+        ConfigImpl.trace("Nothing else in the merge stack, replacing with null", context.depth)
+        nil
+      else
+        # generate a new merge stack from only the remaining items
+        merged = nil
+        sub_stack.each do |v|
+          if merged.nil?
+            merged = v
+          else
+            merged = merged.with_fallback(v)
+          end
+        end
+        merged
+      end
+    end
+
+  end
+
   def resolve_status
     Hocon::Impl::ResolveStatus::UNRESOLVED
   end
 
+  def replace_child(child, replacement)
+    new_stack = replace_child_in_list(stack, child, replacement)
+    if new_stack.nil?
+      nil
+    else
+      self.class.new(origin, new_stack)
+    end
+  end
+
+  def has_descendant(descendant)
+    has_descendant_in_list(stack, descendant)
```

Comment:
`has_descendant_in_list` should be static. Also, it looks like it's unimplemented. I'm fine with punting on it until we run into a test failure though.